lps22: add a component

nagisa opened the pull request:
# What does this implement/fix?

LPS22HB/HH is a barometric pressure sensor. It has pretty high accuracy, is somewhat less prone to hallucinating weird measurements compared to competing chips and is reasonably widely available for purchase.

This adds a very basic one-shot based implementation for interacting with this sensor over an I2C bus. The chip also supports communication over SPI, but I don't have the means to test such configuration (for my PCB is baked for I2C) as well as measuring the conditions in a automatic mode. These are left as future extensions to the basic implementation seen here.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** N/S

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#4340

## Test Environment

- [x] ESP32
  * tested with real hw
- [x] ESP32 IDF
  * tested with real hw
- [x] ESP8266
- [x] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
sensor:
  - platform: lps22
    temperature:
      name: "LPS22 Temperature"
    pressure:
      name: "LPS22 Pressure"
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

nagisa committed:
lps22: add a component

LPS22HB/HH is a barometric pressure sensor. It has pretty high accuracy,
is somewhat less prone to measuring weird stuff compared to competing
chips and is reasonably widely available for purchase.

This adds a very basic one-shot based implementation for interacting
with this sensor over an I2C bus. The chip also supports communication
over SPI, but I don't have the means to test such configuration (for my
PCB is baked for I2C) as well as measuring the conditions in a automatic
mode. These are left as future extensions to the basic implementation
seen here.

nagisa committed:
Merge branch 'dev' into lps22

jesserockz commented on esphome/components/lps22/lps22.cpp:
```diff
@@ -0,0 +1,66 @@
+void LPS22Component::dump_config() { LOG_I2C_DEVICE(this); }
+
+void LPS22Component::update() {
+  uint8_t value = 0x00;
+  this->read_register(WHO_AM_I, &value, 1);
+  if (value != LPS22HB_ID && value != LPS22HH_ID) {
+    ESP_LOGW(TAG, "device IDs as %02x, which isn't a known LPS22HB or LPS22HH ID", value);
+    return;
+  }
```
This should be in `setup()` and also call `this->mark_failed()` if the device is wrong.

nagisa replied:
Seems to work well. Thank you!

jesserockz commented on esphome/components/lps22/lps22.cpp:
```diff
@@ -0,0 +1,66 @@
+static constexpr uint8_t REF_P_XL = 0x28;
+static constexpr uint8_t READ_ATTEMPTS = 10;
+static constexpr uint8_t READ_INTERVAL = 5;
+static constexpr float PRESSURE_SCALE = 1. / 4096.;
+static constexpr float TEMPERATURE_SCALE = 0.01;
```
```suggestion
static constexpr float PRESSURE_SCALE = 1.0f / 4096.0f;
static constexpr float TEMPERATURE_SCALE = 0.01f;
```

jesserockz commented on esphome/components/lps22/lps22.cpp:
```diff
@@ -0,0 +1,66 @@
+    return RetryResult::RETRY;
+  }
+
+  uint8_t t_buf[2]{0};
+  this->read_register(TEMP_L, t_buf, 2);
+  float temp =
+      TEMPERATURE_SCALE * static_cast<float>(static_cast<int16_t>(t_buf[1]) << 8 | static_cast<int16_t>(t_buf[0]));
+  this->temperature_sensor_->publish_state(temp);
```
```suggestion
  if(this->temperature_sensor_ != nullptr) {
    uint8_t t_buf[2]{0};
    this->read_register(TEMP_L, t_buf, 2);
    float temp =
        TEMPERATURE_SCALE * static_cast<float>(encode_uint16(t_buf[1], t_buf[0]));
    this->temperature_sensor_->publish_state(temp);
  }
```

jesserockz commented on esphome/components/lps22/lps22.cpp:
```diff
@@ -0,0 +1,66 @@
+  float temp =
+      TEMPERATURE_SCALE * static_cast<float>(static_cast<int16_t>(t_buf[1]) << 8 | static_cast<int16_t>(t_buf[0]));
+  this->temperature_sensor_->publish_state(temp);
+  uint8_t p_buf[3]{0};
+  this->read_register(PRES_OUT_XL, p_buf, 3);
+  uint32_t p_lsb =
+      static_cast<uint32_t>(p_buf[2]) << 16 | static_cast<uint32_t>(p_buf[1]) << 8 | static_cast<uint32_t>(p_buf[0]);
+  this->pressure_sensor_->publish_state(PRESSURE_SCALE * static_cast<float>(p_lsb));
```
```suggestion
  if (this->pressure_sensor_ != nullptr) {
    uint8_t p_buf[3]{0};
    this->read_register(PRES_OUT_XL, p_buf, 3);
    uint32_t p_lsb = encode_uint24(p_buf[2], p_buf[1], p_buf[0]);
    this->pressure_sensor_->publish_state(PRESSURE_SCALE * static_cast<float>(p_lsb));
  }
```

jesserockz commented on esphome/components/lps22/lps22.cpp:
```diff
@@ -0,0 +1,66 @@
+
+void LPS22Component::setup() {}
+
+void LPS22Component::dump_config() { LOG_I2C_DEVICE(this); }
```
```suggestion
void LPS22Component::dump_config() {
  ESP_LOGCONFIG(TAG, "LPS22:");
  LOG_SENSOR("  ", "Temperature", this->temperature_sensor_);
  LOG_SENSOR("  ", "Pressure", this->pressure_sensor_);
  LOG_I2C_DEVICE(this);
  LOG_UPDATE_INTERVAL(this);
}
```

jesserockz commented on esphome/components/lps22/sensor.py:
```diff
@@ -0,0 +1,58 @@
+    await cg.register_component(var, config)
+    await i2c.register_i2c_device(var, config)
+
+    if CONF_TEMPERATURE in config:
+        sens = await sensor.new_sensor(config[CONF_TEMPERATURE])
```
```suggestion
    if temperature_config := config.get(CONF_TEMPERATURE):
        sens = await sensor.new_sensor(temperature_config)
```

jesserockz commented on esphome/components/lps22/sensor.py:
```diff
@@ -0,0 +1,58 @@
+        sens = await sensor.new_sensor(config[CONF_TEMPERATURE])
+        cg.add(var.set_temperature_sensor(sens))
+
+    if CONF_PRESSURE in config:
+        sens = await sensor.new_sensor(config[CONF_PRESSURE])
```
```suggestion
    if pressure_config := config.get(CONF_PRESSURE):
        sens = await sensor.new_sensor(pressure_config)
```

jesserockz commented on esphome/components/lps22/lps22.h:
```diff
@@ -0,0 +1,30 @@
+  sensor::Sensor *temperature_sensor_{nullptr};
+  sensor::Sensor *pressure_sensor_{nullptr};
+
+ private:
```
We avoid private when not necessary
```suggestion
```

jesserockz commented on esphome/components/lps22/lps22.h:
```diff
@@ -0,0 +1,30 @@
+
+class LPS22Component : public sensor::Sensor, public PollingComponent, public i2c::I2CDevice {
+ public:
+  void set_temperature_sensor(sensor::Sensor *temperature_sensor) { temperature_sensor_ = temperature_sensor; }
+  void set_pressure_sensor(sensor::Sensor *pressure_sensor) { pressure_sensor_ = pressure_sensor; }
```
```suggestion
  void set_temperature_sensor(sensor::Sensor *temperature_sensor) { this->temperature_sensor_ = temperature_sensor; }
  void set_pressure_sensor(sensor::Sensor *pressure_sensor) { this->pressure_sensor_ = pressure_sensor; }
```

jesserockz commented on esphome/components/lps22/lps22.h:
```diff
@@ -0,0 +1,30 @@
+  void set_temperature_sensor(sensor::Sensor *temperature_sensor) { temperature_sensor_ = temperature_sensor; }
+  void set_pressure_sensor(sensor::Sensor *pressure_sensor) { pressure_sensor_ = pressure_sensor; }
+
+  LPS22Component() : PollingComponent(15000) {}
```
Not needed
```suggestion
```

nagisa committed:
Update esphome/components/lps22/lps22.h

Co-authored-by: Jesse Hills <3060199+jesserockz@users.noreply.github.com>

nagisa committed:
Update esphome/components/lps22/lps22.h

Co-authored-by: Jesse Hills <3060199+jesserockz@users.noreply.github.com>

nagisa committed:
Update esphome/components/lps22/lps22.h

Co-authored-by: Jesse Hills <3060199+jesserockz@users.noreply.github.com>

nagisa committed:
Update esphome/components/lps22/sensor.py

Co-authored-by: Jesse Hills <3060199+jesserockz@users.noreply.github.com>

nagisa committed:
Update esphome/components/lps22/lps22.cpp

Co-authored-by: Jesse Hills <3060199+jesserockz@users.noreply.github.com>

nagisa committed:
Update esphome/components/lps22/sensor.py

Co-authored-by: Jesse Hills <3060199+jesserockz@users.noreply.github.com>

nagisa committed:
Update esphome/components/lps22/lps22.cpp

Co-authored-by: Jesse Hills <3060199+jesserockz@users.noreply.github.com>

nagisa committed:
Update esphome/components/lps22/lps22.cpp

Co-authored-by: Jesse Hills <3060199+jesserockz@users.noreply.github.com>

nagisa committed:
Update esphome/components/lps22/lps22.cpp

Co-authored-by: Jesse Hills <3060199+jesserockz@users.noreply.github.com>

nagisa committed:
lps22: move device ID check to setup()

nagisa commented:
Thank you very much for the thorough review. The code seems much nicer now!

nagisa committed:
lps22: fix interpretation of temperature bytes

Temperatures are represented as signed int16_t, allowing the sensor to
represent negative temperatures (in °C.) I only noticed just now because
it has been getting somewhat cold recently :)

jesserockz commented:
> Thank you very much for the thorough review. The code seems much nicer now! Let me know if you want me to squash commits or anything else.

https://esphome.io/guides/contributing#force-push

nagisa committed:
Merge branch 'dev' into lps22

kbx81 committed:
Merge branch 'dev' into lps22

kbx81 committed:
Merge branch 'dev' into lps22
